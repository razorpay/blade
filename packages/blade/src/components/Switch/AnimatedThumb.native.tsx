/* eslint-disable react-hooks/exhaustive-deps */
import styled from 'styled-components/native';
import getIn from 'lodash/get';
import React from 'react';
import isNumber from 'lodash/isNumber';
import Animated, {
  interpolate,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';

import { switchColors, switchMotion, switchSizes } from './switchTokens';
import type { AnimatedThumbProps } from './types';
import { useBreakpoint } from '~utils';
import { useTheme } from '~components/BladeProvider';
import { makeBorderSize } from '~utils/makeBorderSize';

const StyledAnimatedThumb = styled(Animated.View)<{ isDisabled?: boolean }>(
  ({ theme, isDisabled }) => {
    const variant = isDisabled ? 'disabled' : 'default';
    const backgroundColor = getIn(theme, switchColors.thumb[variant].background);

    return {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      flexShrink: 0,
      width: '100%',
      height: '100%',
      borderRadius: makeBorderSize(theme.border.radius.max),
      backgroundColor,
      position: 'absolute',
    };
  },
);

const AnimatedThumb = ({
  isChecked,
  isDisabled,
  size = 'medium',
  children,
  isPressed,
}: AnimatedThumbProps): React.ReactElement => {
  const { theme } = useTheme();
  const { matchedDeviceType } = useBreakpoint({ breakpoints: theme.breakpoints });
  const sharedLeft = useSharedValue(isChecked ? 1 : 0);
  const sharedWidth = useSharedValue(isPressed ? 1 : 0);
  const sharedShouldShiftOffset = useSharedValue(Boolean(isChecked && isPressed));

  const easing = getIn(theme, switchMotion.easing.thumb);
  const duration = getIn(theme, switchMotion.duration.thumb);
  const thumbWidth = switchSizes.thumb[matchedDeviceType][size].width;
  const finalWidth = isNumber(thumbWidth) ? thumbWidth : getIn(theme, thumbWidth);

  React.useEffect(() => {
    sharedLeft.value = withTiming(isChecked ? 1 : 0, {
      duration,
      easing,
    });
  }, [isChecked]);

  React.useEffect(() => {
    sharedWidth.value = withTiming(isPressed ? 1 : 0, {
      duration,
      easing,
    });
  }, [isPressed]);

  React.useEffect(() => {
    sharedShouldShiftOffset.value = Boolean(isChecked && isPressed);
  }, [isChecked, isPressed]);

  const thumbAnimation = useAnimatedStyle(() => {
    return {
      width: interpolate(
        sharedWidth.value,
        [0, 1],
        // scale thumb by 25%, 1.25 comes from motion guidelines
        [finalWidth, finalWidth * 1.25],
      ),
      left: withTiming(
        // While on checked state, shift the thumb 25% to left because
        // We elongate the width 25% to right.
        sharedShouldShiftOffset.value ? finalWidth * -0.25 : 0,
        { easing, duration },
      ),
      transform: [
        {
          translateX: interpolate(sharedLeft.value, [0, 1], [0, finalWidth]),
        },
      ],
    };
  }, []);

  return (
    <StyledAnimatedThumb style={thumbAnimation} isDisabled={isDisabled}>
      {children}
    </StyledAnimatedThumb>
  );
};

export { AnimatedThumb };
