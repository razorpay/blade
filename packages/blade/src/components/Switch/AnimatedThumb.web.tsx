import styled, { css } from 'styled-components';
import { switchColors, switchMotion } from './switchTokens';
import type { AnimatedThumbProps } from './types';
import getIn from '~utils/lodashButBetter/get';
import { BaseBox } from '~components/Box/BaseBox';
import { makeMotionTime } from '~utils/makeMotionTime';
import { makeBorderSize } from '~utils/makeBorderSize';

const AnimatedThumb = styled(BaseBox)<AnimatedThumbProps>(
  ({ theme, isChecked, isDisabled, isPressed }) => {
    const variant = isDisabled ? 'disabled' : 'default';
    const backgroundColor = getIn(theme, switchColors.thumb[variant].background);
    const duration = makeMotionTime(getIn(theme, switchMotion.duration.thumb));
    // offset the thumb x% left because we change the width and width changes from the center origin of thumb
    const offset = isChecked ? '-39%' : '12.5%';
    return css`
      display: flex;
      align-items: center;
      justify-content: center;
      flex-shrink: 0;
      height: 100%;
      position: relative;
      will-change: transform, left;
      width: ${isPressed ? '125%' : '100%'};
      left: ${isPressed ? offset : '0%'};
      transform: translateX(${isChecked ? '100%' : '0%'});
      transition: ${duration};
      border-radius: ${makeBorderSize(theme.border.radius.max)};
      animation-duration: ${duration};
      background-color: ${backgroundColor};
    `;
  },
);

export { AnimatedThumb };
