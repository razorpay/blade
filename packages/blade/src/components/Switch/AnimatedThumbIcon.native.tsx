import React from 'react';
import Animated, {
  interpolate,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';
import { switchMotion } from './switchTokens';
import getIn from '~utils/lodashButBetter/get';
import { useTheme } from '~components/BladeProvider';
import type { SvgProps } from '~components/Icons/_Svg/Svg/types';
import Svg from '~components/Icons/_Svg';

const AnimatedThumbIcon = ({
  children,
  isChecked,
  width,
  height,
  viewBox,
  fill,
}: {
  children: React.ReactNode;
  isChecked: boolean;
} & SvgProps): React.ReactElement => {
  const { theme } = useTheme();
  const opacity = useSharedValue(isChecked ? 1 : 0);

  const easing = getIn(theme, switchMotion.easing.thumbIcon);
  const duration = getIn(theme, switchMotion.duration.thumbIcon);

  React.useEffect(() => {
    opacity.value = withTiming(isChecked ? 1 : 0, {
      duration,
      easing,
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isChecked]);

  const opacityStyle = useAnimatedStyle(() => {
    return {
      opacity: interpolate(opacity.value, [0, 0.2, 1], [0, 0, 1]),
    };
  }, []);

  return (
    <Animated.View style={opacityStyle}>
      <Svg width={width} height={height} viewBox={viewBox} fill={fill}>
        {children}
      </Svg>
    </Animated.View>
  );
};

export { AnimatedThumbIcon };
