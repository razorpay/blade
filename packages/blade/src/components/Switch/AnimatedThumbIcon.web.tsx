import getIn from 'lodash/get';
import styled from 'styled-components';
import { switchMotion } from './switchTokens';
import Svg from '~components/Icons/_Svg';
import { makeMotionTime } from '~utils/makeMotionTime';

const AnimatedThumbIcon = styled(Svg)<{ isChecked?: boolean }>(({ isChecked, theme }) => {
  const easing = getIn(theme, switchMotion.easing.thumbIcon);
  const duration = getIn(theme, switchMotion.duration.thumbIcon);

  return {
    transitionDuration: `${makeMotionTime(duration)}`,
    transitionTimingFunction: easing,
    transitionDelay: isChecked ? `${makeMotionTime(theme.motion.delay['2xshort'])}` : `0ms`,
    opacity: isChecked ? 1 : 0,
  };
});

export { AnimatedThumbIcon };
