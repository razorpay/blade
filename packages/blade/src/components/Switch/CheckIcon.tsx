import isNumber from 'lodash/isNumber';
import styled from 'styled-components';
import type { SwitchProps } from './types';
import { useTheme } from '~components/BladeProvider';
import { getIn, makeMotionTime, makeSize, makeSpace, useBreakpoint } from '~utils';
import size from '~tokens/global/size';
import Svg, { Path } from '~components/Icons/_Svg';

const switchIconSize = {
  desktop: {
    small: size[6],
    medium: 'spacing.3',
  },
  mobile: {
    small: 'spacing.3',
    medium: size[10],
  },
} as const;

const AnimatedSvg = styled(Svg)<{ isChecked?: boolean }>(({ isChecked, theme }) => {
  return {
    transitionDuration: `${makeMotionTime(theme.motion.duration.xquick)}`,
    transitionTimingFunction: theme.motion.easing.standard.effective as string,
    transitionDelay: isChecked ? `${makeMotionTime(theme.motion.delay['2xshort'])}` : `0ms`,
    opacity: isChecked ? 1 : 0,
  };
});

const CheckIcon = ({
  isChecked,
  isDisabled,
  size = 'medium',
}: Pick<SwitchProps, 'isChecked' | 'isDisabled' | 'size'>): React.ReactElement => {
  const { theme } = useTheme();
  const { matchedDeviceType } = useBreakpoint({ breakpoints: theme.breakpoints });

  const width = switchIconSize[matchedDeviceType][size];
  const height = switchIconSize[matchedDeviceType][size];
  const finalWidth = isNumber(width) ? makeSize(width) : makeSpace(getIn(theme, width));
  const finalHeight = isNumber(height) ? makeSize(height) : makeSpace(getIn(theme, height));

  return (
    <AnimatedSvg
      isChecked={isChecked}
      width={finalWidth}
      height={finalHeight}
      viewBox="0 0 11 8"
      fill="none"
    >
      <Path
        fillRule="evenodd"
        clipRule="evenodd"
        d="M8.81891 0.546661C9.12722 0.238352 9.62709 0.238353 9.9354 0.546661C10.2437 0.85497 10.2437 1.35484 9.9354 1.66315L4.14592 7.45262C3.83761 7.76093 3.33775 7.76093 3.02944 7.45262L0.397858 4.82104C0.0895488 4.51273 0.0895488 4.01286 0.397857 3.70456C0.706166 3.39625 1.20603 3.39625 1.51434 3.70456L3.58768 5.77789L8.81891 0.546661Z"
        fill={
          isDisabled
            ? theme.colors.surface.text.placeholder.lowContrast
            : theme.colors.brand.gray[200].highContrast
        }
      />
    </AnimatedSvg>
  );
};

export { CheckIcon };
