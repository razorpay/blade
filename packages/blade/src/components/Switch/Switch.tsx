/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/prefer-ts-expect-error */
import React from 'react';
import { ThumbIcon } from './ThumbIcon';
import type { SwitchProps } from './types';
import { switchHoverTokens } from './switchTokens';
import { Thumb } from './Thumb';
import { AnimatedThumb } from './AnimatedThumb';
import { SwitchTrack } from './SwitchTrack';
import { useTheme } from '~components/BladeProvider';
import { BaseBox } from '~components/Box/BaseBox';
import { useCheckbox } from '~components/Checkbox/useCheckbox';
import { SelectorInput } from '~components/Form/Selector/SelectorInput';
import { useBreakpoint } from '~utils';
import { metaAttribute, MetaConstants } from '~utils/metaAttribute';
import { SelectorLabel } from '~components/Form/Selector/SelectorLabel';
import type { BladeElementRef } from '~utils/types';
import { makeAccessible } from '~utils/makeAccessible';
import { assignWithoutSideEffects } from '~utils/assignWithoutSideEffects';

const _Switch: React.ForwardRefRenderFunction<BladeElementRef, SwitchProps> = (
  {
    defaultChecked,
    isChecked,
    isDisabled,
    name,
    onChange,
    size = 'medium',
    value,
    accessibilityLabel,
    testID,
  },
  ref,
): React.ReactElement => {
  const { theme } = useTheme();
  const { matchedDeviceType } = useBreakpoint({ breakpoints: theme.breakpoints });
  const [isPressed, setIsPressed] = React.useState(false);
  const { state, inputProps } = useCheckbox({
    role: 'switch',
    defaultChecked,
    isChecked,
    isIndeterminate: false,
    hasError: false,
    hasHelperText: false,
    isDisabled,
    isRequired: false,
    name,
    value,
    onChange,
  });

  const handlePointerPressedIn = React.useCallback(() => {
    if (isDisabled) return;
    setIsPressed(true);
  }, [isDisabled]);

  const handlePointerPressedOut = React.useCallback(() => {
    if (isDisabled) return;
    setIsPressed(false);
  }, [isDisabled]);

  const handleKeyboardPressedIn = React.useCallback(
    (e: React.KeyboardEvent) => {
      if (isDisabled) return;
      if (e.key === ' ') {
        setIsPressed(true);
      }
    },
    [isDisabled],
  );

  const handleKeyboardPressedOut = React.useCallback(
    (e: React.KeyboardEvent) => {
      if (isDisabled) return;
      if (e.key === ' ') {
        setIsPressed(false);
      }
    },
    [isDisabled],
  );

  return (
    <BaseBox
      {...metaAttribute({ testID, name: MetaConstants.Switch })}
      // @ts-ignore on rn inline-block is not valid but type here will be `flex | inline-block`
      display={state.isReactNative ? 'flex' : 'inline-block'}
    >
      <SelectorLabel
        componentName={MetaConstants.SwitchLabel}
        onTouchStart={handlePointerPressedIn}
        onTouchEnd={handlePointerPressedOut}
        onMouseDown={handlePointerPressedIn}
        onMouseUp={handlePointerPressedOut}
        onMouseOut={handlePointerPressedOut}
        onKeyDown={handleKeyboardPressedIn}
        onKeyUp={handleKeyboardPressedOut}
        style={{ cursor: 'pointer' }}
        inputProps={
          state.isReactNative
            ? // accessibility label for react-native needs to be added
              // since there is no text children inside Switch
              { ...inputProps, ...makeAccessible({ label: accessibilityLabel }) }
            : {}
        }
      >
        <SelectorInput
          hoverTokens={switchHoverTokens}
          ref={ref}
          isChecked={state.isChecked}
          isDisabled={isDisabled}
          hasError={false}
          inputProps={inputProps}
          accessibilityLabel={accessibilityLabel}
        />
        <SwitchTrack
          size={size}
          deviceType={matchedDeviceType}
          isDisabled={Boolean(isDisabled)}
          isChecked={state.isChecked}
          {...makeAccessible({ hidden: true })}
        >
          <Thumb size={size} deviceType={matchedDeviceType} isChecked={state.isChecked}>
            <AnimatedThumb
              isPressed={isPressed}
              isChecked={state.isChecked}
              isDisabled={isDisabled}
              size={size}
            >
              <ThumbIcon isChecked={state.isChecked} isDisabled={isDisabled} size={size} />
            </AnimatedThumb>
          </Thumb>
        </SwitchTrack>
      </SelectorLabel>
    </BaseBox>
  );
};

const Switch = assignWithoutSideEffects(React.forwardRef(_Switch), {
  displayName: 'Switch',
});

export { Switch };
