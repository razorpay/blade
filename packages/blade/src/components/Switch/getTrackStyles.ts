import type { CSSObject } from 'styled-components';
import type { TrackProps } from './types';
import { switchColors, switchSizes } from './switchTokens';
import getIn from '~utils/lodashButBetter/get';
import isNumber from '~utils/lodashButBetter/isNumber';
import { isReactNative } from '~utils';
import type { Theme } from '~components/BladeProvider';
import { makeSize } from '~utils/makeSize';
import { makeSpace } from '~utils/makeSpace';
import { makeMotionTime } from '~utils/makeMotionTime';

const getTrackStyles = ({
  isChecked,
  isDisabled,
  deviceType,
  size,
  theme,
}: TrackProps & {
  theme: Theme;
}): CSSObject => {
  let variant: 'default' | 'disabled' = 'default';
  if (isDisabled) variant = 'disabled';
  const checked = isChecked ? 'checked' : 'unchecked';
  const background = switchColors.track[variant].background[checked];
  const backgroundColor = getIn(theme, background);

  const widthToken = switchSizes.track[deviceType][size].width;
  const heightToken = switchSizes.track[deviceType][size].height;
  const width = isNumber(widthToken) ? makeSize(widthToken) : makeSpace(getIn(theme, widthToken));
  const height = isNumber(heightToken)
    ? makeSize(heightToken)
    : makeSpace(getIn(theme, heightToken));

  return {
    pointerEvents: 'none',
    position: 'relative',
    display: 'flex',
    alignItems: 'center',
    margin: makeSpace(theme.spacing[1]),
    padding: makeSpace(theme.spacing[1]),
    width,
    height,
    borderRadius: makeSize(theme.border.radius.max),
    backgroundColor,
    transitionTimingFunction: `${theme.motion.easing.standard.effective}`,
    transitionDuration: isReactNative()
      ? undefined
      : `${makeMotionTime(theme.motion.duration['2xquick'])}`,
  };
};

export { getTrackStyles };
