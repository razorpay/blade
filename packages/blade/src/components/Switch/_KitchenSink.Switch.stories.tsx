import { composeStories } from '@storybook/react';
import * as switchStories from './Switch.stories';
import { Box } from '~components/Box';
import { Heading } from '~components/Typography';

const allStories = Object.values(composeStories(switchStories));

export const Switch = (): JSX.Element => {
  return (
    <Box display="flex" flexDirection="column" gap="spacing.4">
      {allStories.map((Story) => {
        return (
          <>
            <Heading>{Story.storyName}</Heading>
            <Story />
          </>
        );
      })}
    </Box>
  );
};

export default {
  title: 'Components/KitchenSink/Switch',
  component: Switch,
  parameters: {
    // enable Chromatic's snapshotting only for kitchensink
    chromatic: { disableSnapshot: false },
    options: { showPanel: false },
  },
};
