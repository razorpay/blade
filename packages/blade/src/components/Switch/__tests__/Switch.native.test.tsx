/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { fireEvent } from '@testing-library/react-native';
import React from 'react';
import { Text } from 'react-native';
import { Switch } from '../Switch';
import renderWithTheme from '~utils/testing/renderWithTheme.native';

describe('<Switch />', () => {
  it('should render switch with label', () => {
    const name = 'Toggle Darkmode';
    const { toJSON, getByRole } = renderWithTheme(<Switch accessibilityLabel={name} />);
    expect(toJSON()).toMatchSnapshot();
    expect(getByRole('switch')).toBeDefined();
  });

  it('should set disabled state with isDisabled', () => {
    const name = 'Toggle Darkmode';
    const { toJSON, queryByA11yState } = renderWithTheme(
      <Switch isDisabled accessibilityLabel={name} />,
    );
    expect(toJSON()).toMatchSnapshot();
    const switchElement = queryByA11yState({ disabled: true });
    expect(switchElement).toBeTruthy();
    expect(switchElement?.props?.accessibilityState).toStrictEqual({
      checked: false,
      disabled: true,
    });
  });

  it('should set defaultChecked', () => {
    const name = 'Toggle Darkmode';
    const { getByRole } = renderWithTheme(<Switch defaultChecked accessibilityLabel={name} />);
    const switchElement = getByRole('switch');
    expect(switchElement.props.accessibilityState.checked).toBeTruthy();
  });

  it('should support isChecked prop', () => {
    const name = 'Toggle Darkmode';
    const { getByRole } = renderWithTheme(<Switch isChecked accessibilityLabel={name} />);
    const switchElement = getByRole('switch');

    expect(switchElement.props.accessibilityState.checked).toBeTruthy();
    // should not toggle
    fireEvent.press(switchElement);
    expect(switchElement.props.accessibilityState.checked).toBeTruthy();
  });

  test('user should be able to toggle switch', () => {
    const name = 'Toggle Darkmode';
    const { getByRole } = renderWithTheme(<Switch accessibilityLabel={name} />);
    const switchElement = getByRole('switch');

    expect(switchElement.props.accessibilityState.checked).toBeFalsy();
    fireEvent.press(switchElement);
    expect(switchElement.props.accessibilityState.checked).toBeTruthy();
    fireEvent.press(switchElement);
    expect(switchElement.props.accessibilityState.checked).toBeFalsy();
  });

  it('should support uncontrolled state', () => {
    const checkFn = jest.fn();
    const eventData = {
      nativeEvent: {
        pageX: 20,
        pageY: 30,
      },
    };
    const name = 'Toggle Darkmode';
    const { getByRole } = renderWithTheme(
      <Switch defaultChecked={true} onChange={checkFn} accessibilityLabel={name} />,
    );
    const switchElement = getByRole('switch');

    expect(switchElement.props.accessibilityState.checked).toBeTruthy();
    expect(checkFn).not.toBeCalled();
    fireEvent.press(switchElement, eventData);
    expect(switchElement.props.accessibilityState.checked).toBeFalsy();
    expect(checkFn).toBeCalledWith(
      expect.objectContaining({
        isChecked: false,
        value: undefined,
        event: eventData,
      }),
    );
    fireEvent.press(switchElement, eventData);
    expect(switchElement.props.accessibilityState.checked).toBeTruthy();
    expect(checkFn).toBeCalledWith(
      expect.objectContaining({
        isChecked: true,
        value: undefined,
        event: eventData,
      }),
    );
  });

  it('should support controlled state', () => {
    const name = 'Toggle Darkmode';
    const Example = () => {
      const [checked, setChecked] = React.useState(false);
      return (
        <>
          <Switch
            isChecked={checked}
            onChange={({ isChecked }) => setChecked(isChecked)}
            accessibilityLabel={name}
          />
          <Text>{checked ? 'checked' : 'unchecked'}</Text>
        </>
      );
    };
    const { getByText, getByRole } = renderWithTheme(<Example />);
    const switchElement = getByRole('switch');

    expect(getByText('unchecked')).toBeTruthy();
    fireEvent.press(switchElement);
    expect(getByText('checked')).toBeTruthy();
    fireEvent.press(switchElement);
    expect(getByText('unchecked')).toBeTruthy();
  });

  it('should expose native element methods via ref', () => {
    let refValue = null;
    const Example = (): React.ReactElement => {
      const ref = React.useRef<HTMLInputElement>(null);
      return (
        <Switch
          accessibilityLabel="Toggle darkmode"
          ref={(value) => {
            console.log(value);
            // @ts-expect-error
            ref.current = value;
            refValue = value;
          }}
        />
      );
    };

    renderWithTheme(<Example />);
    expect(refValue).toBeNull();
  });

  it('should accept testID', () => {
    const name = 'Toggle Darkmode';
    const { getByTestId } = renderWithTheme(
      <Switch testID="switchElement-test" accessibilityLabel={name} />,
    );
    expect(getByTestId('switchElement-test')).toBeTruthy();
  });
});
