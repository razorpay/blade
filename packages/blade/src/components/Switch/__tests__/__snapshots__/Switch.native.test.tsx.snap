// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> should render switch with label 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    data-blade-component="switch"
    display="flex"
    style={
      [
        {
          "display": "flex",
        },
      ]
    }
  >
    <View
      accessibilityInvalid={false}
      accessibilityLabel="Toggle Darkmode"
      accessibilityRequired={false}
      accessibilityRole="switch"
      accessibilityState={
        {
          "busy": undefined,
          "checked": false,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "display": "flex",
            "flexDirection": "row",
            "marginBottom": 2,
            "marginTop": 2,
          },
        ]
      }
    >
      <View
        accessibilityElementsHidden={true}
        deviceType="mobile"
        importantForAccessibility="no-hide-descendants"
        isChecked={false}
        isDisabled={false}
        size="medium"
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "hsla(211, 20%, 52%, 0.12)",
              "borderRadius": 9999,
              "display": "flex",
              "height": 24,
              "marginBottom": 2,
              "marginLeft": 2,
              "marginRight": 2,
              "marginTop": 2,
              "paddingBottom": 2,
              "paddingLeft": 2,
              "paddingRight": 2,
              "paddingTop": 2,
              "pointerEvents": "none",
              "position": "relative",
              "transitionTimingFunction": "0.3 0 0.2 1",
              "width": 44,
            },
            {
              "backgroundColor": undefined,
            },
          ]
        }
      >
        <View
          data-blade-component="base-box"
          deviceType="mobile"
          isChecked={false}
          size="medium"
          style={
            [
              {},
              {
                "alignItems": "center",
                "display": "flex",
                "height": 20,
                "justifyContent": "center",
                "left": 0,
                "marginBottom": 2,
                "marginLeft": 2,
                "marginRight": 2,
                "marginTop": 2,
                "position": "absolute",
                "width": 20,
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "hsla(0, 0%, 100%, 1)",
                  "borderRadius": 9999,
                  "display": "flex",
                  "flexShrink": 0,
                  "height": "100%",
                  "justifyContent": "center",
                  "position": "absolute",
                  "width": "100%",
                },
                {
                  "left": 0,
                  "transform": [
                    {
                      "translateX": undefined,
                    },
                  ],
                  "width": undefined,
                },
              ]
            }
          >
            <View
              style={
                {
                  "opacity": undefined,
                }
              }
            >
              <RNSVGSvgView
                accessibilityElementsHidden={true}
                align="xMidYMid"
                bbHeight="10px"
                bbWidth="10px"
                data-blade-component="icon"
                fill="none"
                focusable={false}
                height="10px"
                importantForAccessibility="no-hide-descendants"
                meetOrSlice={0}
                minX={0}
                minY={0}
                style={
                  [
                    {
                      "backgroundColor": "transparent",
                      "borderWidth": 0,
                    },
                    [
                      {},
                    ],
                    {
                      "flex": 0,
                      "height": 10,
                      "width": 10,
                    },
                  ]
                }
                vbHeight={8}
                vbWidth={11}
                width="10px"
              >
                <RNSVGGroup
                  fill={null}
                  propList={
                    [
                      "fill",
                    ]
                  }
                >
                  <RNSVGPath
                    clipRule={0}
                    d="M8.81891 0.546661C9.12722 0.238352 9.62709 0.238353 9.9354 0.546661C10.2437 0.85497 10.2437 1.35484 9.9354 1.66315L4.14592 7.45262C3.83761 7.76093 3.33775 7.76093 3.02944 7.45262L0.397858 4.82104C0.0895488 4.51273 0.0895488 4.01286 0.397857 3.70456C0.706166 3.39625 1.20603 3.39625 1.51434 3.70456L3.58768 5.77789L8.81891 0.546661Z"
                    fill={
                      {
                        "payload": 4278190080,
                        "type": 0,
                      }
                    }
                    fillRule={0}
                    propList={
                      [
                        "fill",
                        "fillRule",
                      ]
                    }
                  />
                </RNSVGGroup>
              </RNSVGSvgView>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`<Switch /> should set disabled state with isDisabled 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    data-blade-component="switch"
    display="flex"
    style={
      [
        {
          "display": "flex",
        },
      ]
    }
  >
    <View
      accessibilityInvalid={false}
      accessibilityLabel="Toggle Darkmode"
      accessibilityRequired={false}
      accessibilityRole="switch"
      accessibilityState={
        {
          "busy": undefined,
          "checked": false,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "display": "flex",
            "flexDirection": "row",
            "marginBottom": 2,
            "marginTop": 2,
          },
        ]
      }
    >
      <View
        accessibilityElementsHidden={true}
        deviceType="mobile"
        importantForAccessibility="no-hide-descendants"
        isChecked={false}
        isDisabled={true}
        size="medium"
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "hsla(211, 20%, 52%, 0.09)",
              "borderRadius": 9999,
              "display": "flex",
              "height": 24,
              "marginBottom": 2,
              "marginLeft": 2,
              "marginRight": 2,
              "marginTop": 2,
              "paddingBottom": 2,
              "paddingLeft": 2,
              "paddingRight": 2,
              "paddingTop": 2,
              "pointerEvents": "none",
              "position": "relative",
              "transitionTimingFunction": "0.3 0 0.2 1",
              "width": 44,
            },
            {
              "backgroundColor": undefined,
            },
          ]
        }
      >
        <View
          data-blade-component="base-box"
          deviceType="mobile"
          isChecked={false}
          size="medium"
          style={
            [
              {},
              {
                "alignItems": "center",
                "display": "flex",
                "height": 20,
                "justifyContent": "center",
                "left": 0,
                "marginBottom": 2,
                "marginLeft": 2,
                "marginRight": 2,
                "marginTop": 2,
                "position": "absolute",
                "width": 20,
              },
            ]
          }
        >
          <View
            isDisabled={true}
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "hsla(0, 0%, 100%, 0.18)",
                  "borderRadius": 9999,
                  "display": "flex",
                  "flexShrink": 0,
                  "height": "100%",
                  "justifyContent": "center",
                  "position": "absolute",
                  "width": "100%",
                },
                {
                  "left": 0,
                  "transform": [
                    {
                      "translateX": undefined,
                    },
                  ],
                  "width": undefined,
                },
              ]
            }
          >
            <View
              style={
                {
                  "opacity": undefined,
                }
              }
            >
              <RNSVGSvgView
                accessibilityElementsHidden={true}
                align="xMidYMid"
                bbHeight="10px"
                bbWidth="10px"
                data-blade-component="icon"
                fill="none"
                focusable={false}
                height="10px"
                importantForAccessibility="no-hide-descendants"
                meetOrSlice={0}
                minX={0}
                minY={0}
                style={
                  [
                    {
                      "backgroundColor": "transparent",
                      "borderWidth": 0,
                    },
                    [
                      {},
                    ],
                    {
                      "flex": 0,
                      "height": 10,
                      "width": 10,
                    },
                  ]
                }
                vbHeight={8}
                vbWidth={11}
                width="10px"
              >
                <RNSVGGroup
                  fill={null}
                  propList={
                    [
                      "fill",
                    ]
                  }
                >
                  <RNSVGPath
                    clipRule={0}
                    d="M8.81891 0.546661C9.12722 0.238352 9.62709 0.238353 9.9354 0.546661C10.2437 0.85497 10.2437 1.35484 9.9354 1.66315L4.14592 7.45262C3.83761 7.76093 3.33775 7.76093 3.02944 7.45262L0.397858 4.82104C0.0895488 4.51273 0.0895488 4.01286 0.397857 3.70456C0.706166 3.39625 1.20603 3.39625 1.51434 3.70456L3.58768 5.77789L8.81891 0.546661Z"
                    fill={
                      {
                        "payload": 1375731712,
                        "type": 0,
                      }
                    }
                    fillRule={0}
                    propList={
                      [
                        "fill",
                        "fillRule",
                      ]
                    }
                  />
                </RNSVGGroup>
              </RNSVGSvgView>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
